### PR TITLE
Handle refinalization 833

### PR DESCRIPTION
**Pull request by JoeArmani**
Related issue: https://github.com/episphere/connect/issues/833

Related PR: https://github.com/episphere/connectFaas/pull/497
Note: Merge with ConnectFaas PR.

-Update the finalization process to handle re-finalization without resetting 'boxedStatus' property.
-This is needed to handle found stray tubes (when other tubes are already boxed/shipped).

### CODE REVIEW

File: src/events.js [hunk 1]
```diff
@@ -1385,7 +1384,7 @@ const collectionSubmission = async (participantData, biospecimenData, cntd) => {
         if (tubeConceptId && input.required) biospecimenData[tubeConceptId][conceptIds.collection.tube.scannedId] = `${masterID} ${tubeID}`.trim();
     });
 
-    if ((hasError && cntd == true) || hasCntdError) return;
+    if ((hasError && continueToFinalizeScreen == true) || hasCntdError) return;
```
Comment (Gbarra9) [inline src/events.js:1387]: Thanks for renaming `cntd`

File: src/shared.js [hunk 2]
```diff
@@ -3057,4 +3141,105 @@ export const getCurrentDate = () => {
...
-  
\ No newline at end of file
+
+/**
+ * Traverse two objects and compare values. Useful for checking if changes have been made to a data structure or the values within it.
+ * @param {object} originalData - The original object to compare against (deep copy prior to edits).
+ * @param {object} currentData - The current object to compare against after edits (on form submission or similar).
+ * @returns {boolean} - True if the objects are different in shape or values. False if they are the same.
+ */
+export const hasObjectChanged = (originalData, currentData) => {
```
Comment (Gbarra9) [inline src/shared.js:3151]: Nice object comparison checker function

**Review by Gbarra9 — APPROVED** [review]
Looks good

**Review by jhflorey — APPROVED** [review]

**Review by anthonypetersen — APPROVED** [review]

File: src/shared.js (hunk 2 above)
Comment (JoeArmani) [reply src/shared.js:3151]: Thanks!

File: src/events.js (hunk 1 above)
Comment (JoeArmani) [reply src/events.js:1387]: Thanks!

File: src/events.js [hunk 3]
```diff
@@ -1476,45 +1475,102 @@ const collectionSubmission = async (participantData, biospecimenData, cntd) => {
...
+const handleFinalizedCollectionUpdate = async (biospecimenData, participantData, siteTubesList, addedStrayTubes, continueToFinalizeScreen) => {
+    const modalMessage = {
+        title: `Collection ${biospecimenData[conceptIds.collection.id]} is Already Finalized`,
+        body: 'IMPORTANT: This Collection has already been finalized. Click continue if you want to update the collection and re-finalize. Click Cancel to discard changes.',
+    };
 
-    if(baselineVisit && clinicalResearchSetting) {
-        await updateBaselineData(siteTubesList, participantData);
-    }
+    const onCancel = () => { /* Nothing to do here */ };
```
[the remark this answers is not in the document]
Comment (Gbarra9) [reply src/events.js:1518]: @JoeArmani Makes sense to me.

**Pull request merged**